wgpu-core/src/error.rs

```diff
@@ -0,0 +1,321 @@
+                }
+                if let Some(tex) = tex_opt {
+                    let name = crate::gfx_select!(tex => global.texture_label(tex));
+                    ret.push_str(&format_label_line("destination", &name));
+                }
+            }
+            _ => {}
+        };
+        ret
+    }
+}
+
+impl PrettyError for crate::command::PassErrorScope {
+    fn fmt_pretty(&self, global: &Global<IdentityManagerFactory>) -> String {
+        // This error is not in the error chain, only notes are needed
+        match *self {
+            Self::Pass(id) => {
+                let name = crate::gfx_select!(id => global.command_buffer_label(id));
+                format_label_line("command buffer", &name)
+            }
+            Self::SetBindGroup(id) => {
+                let name = crate::gfx_select!(id => global.bind_group_label(id));
+                format_label_line("bind group", &name)
+            }
+            Self::SetPipelineRender(id) => {
+                let name = crate::gfx_select!(id => global.render_pipeline_label(id));
+                format_label_line("render pipeline", &name)
+            }
+            Self::SetPipelineCompute(id) => {
+                let name = crate::gfx_select!(id => global.compute_pipeline_label(id));
+                format_label_line("compute pipeline", &name)
+            }
+            Self::SetVertexBuffer(id) => {
+                let name = crate::gfx_select!(id => global.buffer_label(id));
+                format_label_line("buffer", &name)
+            }
+            Self::SetIndexBuffer(id) => {
+                let name = crate::gfx_select!(id => global.buffer_label(id));
+                format_label_line("buffer", &name)
+            }
+            Self::Draw { pipeline, .. } => {
+                if let Some(id) = pipeline {
+                    let name = crate::gfx_select!(id => global.render_pipeline_label(id));
+                    format_label_line("render pipeline", &name)
+                } else {
+                    String::new()
+                }
+            }
+            Self::Dispatch { pipeline, .. } => {
+                if let Some(id) = pipeline {
+                    let name = crate::gfx_select!(id => global.compute_pipeline_label(id));
+                    format_label_line("compute pipeline", &name)
+                } else {
+                    String::new()
+                }
+            }
+            _ => String::new(),
+        }
+    }
+}
+
+impl PrettyError for ContextError {
+    fn fmt_pretty(&self, _global: &Global<IdentityManagerFactory>) -> String {
+        format_error_line(self.as_display()) + &format_label_line(self.label_key, &self.label)
+    }
+}
+
+pub fn format_error_line(err: &dyn fmt::Display) -> String {
+    format!("    {}\n", err)
+}
+
+pub fn format_note_line(note: &dyn fmt::Display) -> String {
+    format!("      note: {}\n", note)
+}
+
+pub fn format_label_line(label_key: &str, label_value: &str) -> String {
```

Comment:
It would be elegant to make it accept an `fmt::Formatter`, so that it doesn't allocate anything.
Yes, I know the performance doesn't matter here, so this isn't required to be elegant :)

Reply:
I think I attempted that while back when I was initially implementing this, and I couldn't figure it out quite right.
The problem is that you can't really construct a `fmt::Formatter`, so it is only really useful in a context where you already have it, like within `Display`/`Debug`. 

Current call sites want the result as a string, and as I understood it, I would have to wrap the stuff to something that implements Display to get the result, and I didn't want to deal with that.

I suppose it could return `fmt::Arguments`, but I guess it would do some allocation anyway? 

If we happen to get rid of the whole `PrettyError` in favour of the `Display`, then the issue is moot and `fmt::Formatter` is the easiest choice.

Reply:
sorry, it would be generic over `fmt::Write`, not `fmt::Formatter`

Reply:
Oh derp. I should of realized that myself, I implemented this now, it somehow makes a bit more sense now :)